Cpfed 3851 keyboard trap all redhat menu

kelsS opened the pull request:
---
name: Cpfed 3851 keyboard trap all redhat menu
labels: "ready for review"
---

## pfe-navigation

- Add functions to hide mobile menu from keyboard and screen reader when the all red hat mobile menu is open
- Add show mobile menu sr function to postResize
- Add function to send the focus to the search input field when the desktop search toggle is pressed
- Removed some console logs
- Add tab index -1 to places where aria hidden is set to true
- Remove tab index attribute when aria-hidden is set to false (aria hidden needs to be used with tab index -1 to fully work)
- Removed focus styles I added to test with
- Restructured nav so that all the elements were contained in the main nav tag for better screen reader experience
- Removed li items from around the mobile buttons for search and all red hat to improve the screen reader experience
- Added more specificity for hidden attr styles to keep other display settings from overriding them
- Changed li styles to div styles to go along with pfe-nav shadow dom template restructure

## Testing 
- Navigate the desktop and mobile menu with your keyboard using the tab key
- Navigate the desktop and mobile menu with your prefered screen reader using the tab key
- Focus on the All Red Hat menu for mobile and the search toggle for desktop

### Expected behavior: 
- You cannot navigate to the All Red Hat menu without activating the mobile All Red Hat button and you cannot navigate back to the main menu without activating the back to menu button. 
- When tabbing out of the menu it should close all open dropdowns. 
- When activating the search toggle on desktop it should move focus to the search field.

### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

wesruv commented on elements/pfe-navigation/mock/site-switcher.html:
```diff
@@ -66,7 +66,7 @@ <h3>Red Hat Partner Connect</h3>
                       <!-- -->
 
-                      <hr>
+                      <hr aria-hidden="true">
```
Iiinteresting, what do these things do for a screen reader?

wesruv replied:
these things === `hr`

kelsS replied:
@wesruv the screen reader says "horizontal splitter" but with the aria hidden set to true the screen readers just ignores hr tags completely which is why it is always best to do those lines with css rather than with the <hr> tag wherever possible.

wesruv commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -1387,6 +1455,120 @@ class PfeNavigation extends PFElement {
+   * set to the light dom search input field so focus is in the correct place for screen readers and keyboards
+   */
+  _searchFieldFocusHandler() {
+    const searchInput = document.querySelector(".pfe-navigation__search  input");
```
Nice, think we should make our selector more specific. e.g.
```
.pfe-navigation__search input[type='text'],
.pfe-navigation__search input[type='search']
```

kelsS replied:
@wesruv made this change

wesruv commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -1387,6 +1455,120 @@ class PfeNavigation extends PFElement {
+   */
+  _searchFieldFocusHandler() {
+    const searchInput = document.querySelector(".pfe-navigation__search  input");
+    searchInput.focus();
```
Should probably wrap this in an `if` statement to make sure it exists, just in case.

kelsS replied:
@wesruv made this change

wesruv commented on elements/pfe-navigation/src/pfe-navigation.scss:
```diff
@@ -481,7 +483,7 @@ See mixin blocks before selectors that contain the selector name and the three l
 @mixin pfe-navigation__back__before--mobile {
   position: absolute;
-  top: 24px;
+  top: 27px;
```
It'd be nice to document this 'magic number' with a comment.

e.g.
```
  top: 27px; // The height of .foo element
```

kelsS replied:
@wesruv made this change

wesruv reviewed (changes requested):
This is awesome work! I get the sense a lot of testing and research went into it.

There are a few issues that I think need to be addressed before we can merge it, but overall looks great.

wesruv commented:
I popped back in to steal some of your code for Primary Details focus issues and realized we could probably simplify some of this implementation.

Instead of dealing with `this._siteSwitcherMobileOnly` and managing the state of it, anything that should run in mobile only can use:
```
if (this.isSecondaryLinksSectionCollapsed()) {
```

That utility function will return true if we're in mobile mode where all red hat does the swipe instead of the dropdown.

kelsS commented:
> This is awesome work! I get the sense a lot of testing and research went into it.
> 
> There are a few issues that I think need to be addressed before we can merge it, but overall looks great.

Thanks for reviewing and giving feedback @wesruv! I am working on refactoring the code and addressing issues, will keep you posted.

kelsS commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -1387,6 +1455,120 @@ class PfeNavigation extends PFElement {
+  /**
+   * Hide main menu from screen readers and keyboard when mobile All Red Hat menu is open
+   */
+  _hideMobileMainMenuSr() {
```
@wesruv I am working to minimize how many individual things need to be selected but to get it down to even less the menu structure would need to be refactored. Should I work on that or minimize as much as I can without restructuring the nav?

kelsS replied:
@wesruv I got the selectors down to three elements without nav restructuring.

kelsS commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -198,6 +207,16 @@ class PfeNavigation extends PFElement {
     }
+
+    // Tab key listener attached to the last focusable element in the component
+    this._lastFocusableNavElement.addEventListener("keydown", this._tabKeyEventListener);
```
@wesruv could we meet and go over this a bit more with the way I have refactored the code?

kelsS commented:
@wesruv would you take a look at those merge conflicts? I am not sure what to keep or delete.

kelsS commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -102,9 +102,14 @@ class PfeNavigation extends PFElement {
-    this._stickyHandler = this._stickyHandler.bind(this);
+    // Get all focusable elements inside of nav
+    this._focusableElements = 'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';
+    this._focusableNavContent = this.shadowRoot.querySelectorAll(this._focusableElements);
```
@wesruv made this change

kelsS commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -102,9 +102,14 @@ class PfeNavigation extends PFElement {
+    this._focusableElements = 'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])';
+    this._focusableNavContent = this.shadowRoot.querySelectorAll(this._focusableElements);
+    // Get the last focusable elements of Nav and All Red Hat sub menu
+    this._lastFocusableNavElement = this._focusableNavContent[this._focusableNavContent.length - 1];
```
@wesruv made this change

kelsS commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -198,6 +207,16 @@ class PfeNavigation extends PFElement {
+    this._lastFocusableNavElement.addEventListener("keydown", this._tabKeyEventListener);
+
+    if (this._siteSwitcherMobileOnly === null) {
+      return;
```
@wesruv made this change

kelsS commented on elements/pfe-navigation/src/pfe-navigation.js:
```diff
@@ -211,7 +230,14 @@ class PfeNavigation extends PFElement {
+    this._lastFocusableNavElement.removeEventListener("keydown", this._tabKeyEventListener);
+
+    if (this._siteSwitcherMobileOnly === null) {
+      return;
```
@wesruv made this change

wesruv reviewed (approved):
Looks good to me, I just have some questions

wesruv commented:
I'll take a look at the conflicts.

wesruv commented:
That was an _ugly_ diff, it was mismatching a bunch of code, but I think I got it @kelsS please double check me:
https://github.com/patternfly/patternfly-elements/pull/1236/commits/14dedae58133cca978fcfe05ec30514ece56f37f

kelsS commented on elements/pfe-navigation/demo/index.html:
```diff
@@ -415,22 +415,22 @@ <h3>
       </ul>
 
-      <li slot="pfe-navigation--custom-links">
+      <div slot="pfe-navigation--custom-links">
```
@wesruv those LIs did not have a wrapper UL in the light dom which was causing aXe and HTML validation errors.

kelsS commented on elements/pfe-navigation/src/pfe-navigation.html:
```diff
@@ -25,15 +25,16 @@
-          <button class="pfe-navigation__all-red-hat-toggle" id="secondary-links__button--all-red-hat">
+        </div>
+        <div>
+          <button class="pfe-navigation__all-red-hat-toggle" id="secondary-links__button--all-red-hat" aria-label="Open All Red Hat menu">
```
@wesruv I considered that but with how the all red hat sub menu was designed I decided to hide the all red hat open button from screen readers when the all red hat menu is open and made a more descriptive aria-label for the back to menu button as well as added an aria only heading that says All Red Hat menu. This made the screen reader experience less confusing. So when the user activates that back to menu button the focus goes back the all red hat menu which announces to them as collapsed. So they only ever open the menu with the all red hat button but never close the menu with it which is why I did not change the label.

kelsS commented on elements/pfe-navigation/src/pfe-navigation.html:
```diff
@@ -1,4 +1,4 @@
-<div id="pfe-navigation__wrapper" class="pfe-navigation__wrapper" aria-label="Main Navigation">
+<nav id="pfe-navigation__wrapper" class="pfe-navigation__wrapper" aria-label="Main Navigation">
```
@wesruv after screen reader testing more I realized that the the nav being there was reading as a secondary nav within the overall pfe-nav which was clunking and kind of confusing and overly verbose to when using a screen reader. It also helps ensure that the screen reader understands that all of the content is a part of the nav not just those menu links. It made the screen reader experience make more sense and easier to understand. It also made sense to me from a semantic point of view since all of the content is a part of the nav, not just those menu items.
